### PR TITLE
Button alignment and size bug fixes

### PR DESCRIPTION
**Pull request by langermank**
**Note**: These are the **non breaking changes** from this [PR](https://github.com/primer/react/pull/2374) 

## Improvements

- Add a `block` prop for full width
- Add `alignContent` prop to align content to [center or start](https://github.com/github/primer/issues/939)
- Use `control` sizing CSS variable values (not using CSS vars just yet)
- Use `height` over padding for more control over sizing

## Visual updates

Addresses a few things missing from this discussion: https://github.com/github/primer/discussions/1016

- Slightly more condensed padding
- Improved `invisible` hover states (will update these colors when new functional colors are addressed)

Closes:
- https://github.com/primer/react/issues/1979
- https://github.com/github/primer/issues/939

### Screenshots

Please refer to the comprehensive Storybook stories- this is just a small sample of some of the changes

| before | after | description |
| -- | -- | -- |
| image | image | trailing action slot alignment |
| image | image | default padding, font-weight |
| image | image | large height |
| icon was blue | image | invisible svg color |
| image | image | trailing action locks right in full width |
| image | image | `invisible` buttons use default text color if visuals are present, and accent blue if no visuals. |

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

**Review by broccolinisoup — APPROVED** [review]
Seems like you addressed all of the comments that we talked on this [PR](https://github.com/primer/react/pull/2374) 😊 Looks good to me!

File: src/Button/types.ts [hunk 1]
```diff
@@ -35,6 +37,10 @@ export type ButtonBaseProps = {
    * Items that are disabled can not be clicked, selected, or navigated through.
    */
   disabled?: boolean
+  /**
+   * Allow button width to fill its container.
+   */
+  block?: boolean
```
[the remark this answers is not in the document]
Comment (langermank) [reply src/Button/types.ts:43]: Me too.. I had it this way originally, but it felt like I was creating unnecessary friction. Since we use block in PVC I think it makes sense to keep it as block here, and maybe in the future do a sweeping change of all `block` props to `fullWidth`

File: src/Button/ButtonBase.tsx [hunk 2]
```diff
@@ -4,32 +4,38 @@ import Box from '../Box'
...
-      return merge.all([getButtonStyles(theme), getSizeStyles(size, variant, false), getVariantStyles(variant, theme)])
-    }, [theme, size, variant])
+      return merge.all([getButtonStyles(theme), getVariantStyles(variant, theme)])
+    }, [theme, variant])
     const sxStyles = useMemo(() => {
       return merge(baseStyles, sxProp as SxProp)
     }, [baseStyles, sxProp])
+    const iconWrapStyles = {
+      display: 'flex',
+      pointerEvents: 'none',
```
[the remark this answers is not in the document]
Comment (langermank) [reply src/Button/ButtonBase.tsx:37]: We had a fun bug once where an svg inside an interactive element (in this case, ActionListItem) was preventing the onClick behavior for a dropdown section. Check this out: https://www.smashingmagazine.com/2018/05/svg-interaction-pointer-events-property/#using-pointer-events-when-mixing-svg-and-html

File: .changeset/angry-spoons-grin.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+"@primer/react": patch
+---
+
+- Add a block prop for full width
+- Add alignContent prop to align content to center or start
+- Use control sizing CSS variable values (not using CSS vars just yet)
+- Use height over padding for more control over sizing
```
[the remark this answers is not in the document]
Comment (langermank) [reply .changeset/angry-spoons-grin.md:8]: It's a bit easier to manage control size with static `height` and `width` _along_ with internal padding. Its especially apparent for IconOnly buttons. For buttons, since we never want a line break using height feels like a safe solution.

File: src/_TextInputInnerAction.tsx [hunk 4]
```diff
@@ -77,7 +82,7 @@ const TextInputAction = forwardRef<HTMLButtonElement, TextInputActionProps>(
     }
 
     return (
-      <Box as="span" className="TextInput-action" margin={1}>
+      <Box as="span" className="TextInput-action" marginLeft={1} marginRight={1}>
```
Comment (langermank) [inline src/_TextInputInnerAction.tsx:85]: This overall height was a bit off here, but I noticed `small` and `large` look too large with a trailing action as well. We could move this to another PR and address that separately as well!

Comment (mperrotti) [reply src/_TextInputInnerAction.tsx:85]: This seems like a small enough change to sneak into this PR 👍

File: src/_TextInputInnerAction.tsx [hunk 5]
```diff
@@ -20,13 +20,18 @@ type TextInputActionProps = Omit<React.ButtonHTMLAttributes<HTMLButtonElement>,
...
 const invisibleButtonStyleOverrides = {
-  color: 'fg.default',
   paddingTop: '2px',
   paddingRight: '4px',
   paddingBottom: '2px',
   paddingLeft: '4px',
   position: 'relative',
 
+  // Button uses both a hash class and selector, so it seems important is required to override here
+  '&[data-component="IconButton"]': {
```
[the remark this answers is not in the document]
Comment (mperrotti) [reply src/_TextInputInnerAction.tsx:30]: Let's do it!

File: .changeset/angry-spoons-grin.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mperrotti) [reply .changeset/angry-spoons-grin.md:8]: Oh I didn't realize we never allowed a line break in buttons.

When the button text doesn't fit on one line, do we: 
- automatically truncate to a single line?
- require the user to handle truncation?
- something else?

File: src/Button/ButtonBase.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mperrotti) [reply src/Button/ButtonBase.tsx:37]: Yea sounds super fun 😂

File: src/Token/TokenBase.tsx [hunk 6]
```diff
@@ -91,28 +89,20 @@ const variants = variant<
       lineHeight: tokenSizes.small,
       paddingLeft: 1,
       paddingRight: 1,
-      // need to explicitly set padding top and bottom to "0" to override default `<button>` element styles
-      // without setting these, the "x" appears vertically mis-aligned
-      paddingTop: 0,
-      paddingBottom: 0,
```
Comment (mperrotti) [inline src/Token/TokenBase.tsx:97]: Nice!

File: src/Button/types.ts [hunk 7]
```diff
@@ -48,7 +54,15 @@ export type ButtonProps = {
    * The trailing icon comes after button content
    */
   trailingIcon?: React.FunctionComponent<React.PropsWithChildren<IconProps>>
+  /**
+   * Trailing action appears to the right of the trailing visual and is always locked to the end
+   */
+  trailingAction?: React.FunctionComponent<React.PropsWithChildren<IconProps>>
```
Comment (mperrotti) [inline src/Button/types.ts:60]: This name is a little confusing. It makes it sound like it's a button within the button. The best alternative I could come up is `endIcon`.

File: src/_TextInputWrapper.tsx [hunk 8]
```diff
@@ -27,16 +27,22 @@ const sizeVariants = variant({
...
       px: 2,
       py: '3px',
       fontSize: 0,
       lineHeight: '20px',
     },
+    medium: {
+      '--inner-action-size': '24px',
+    },
     large: {
+      '--inner-action-size': '28px',
```
Comment (mperrotti) [inline src/_TextInputWrapper.tsx:41]: I assume we picked `'28px'` because it's `primer.control.small.size` in [size primitives](https://primer.style/primitives/spacing#small). Is that right?

File: src/_TextInputWrapper.tsx [hunk 9]
```diff
@@ -27,16 +27,22 @@ const sizeVariants = variant({
...
       px: 2,
       py: '10px',
-      fontSize: 3,
+      lineHeight: '0',
```
Comment (mperrotti) [inline src/_TextInputWrapper.tsx:44]: Removing this line-height doesn't seem to have any affect on leading visuals or trailing visuals/actions. 🔪 it?

```suggestion
```

File: src/_TextInputWrapper.tsx [hunk 10]
```diff
@@ -27,16 +27,22 @@ const sizeVariants = variant({
   prop: 'size',
   variants: {
     small: {
+      '--inner-action-size': '20px',
```
Comment (mperrotti) [inline src/_TextInputWrapper.tsx:30]: Non-blocking suggestion for hard-coded size values: it would be nice if there were comments that identified the design token that the value comes from.

File: src/Button/styles.ts [hunk 11]
```diff
@@ -170,96 +172,140 @@ export const getVariantStyles = (variant: VariantType = 'default', theme?: Theme
...
+  const style = {
+    center: {
+      justifyContent: 'center',
+    },
+    start: {
+      justifyContent: 'flex-start',
+    },
+  }
+  return style[alignContent]
+}
```
Comment (mperrotti) [inline src/Button/styles.ts:311]: Since we only have two possible values that `alignContent` could be, we could simplify this.

```suggestion
export const getAlignContentSize = (alignContent: AlignContent = 'center') => ({
  justifyContent: alignContent === 'center' ? 'center' : 'flex-start' 
})
```

File: src/Button/styles.ts [hunk 12]
```diff
@@ -170,96 +172,140 @@ export const getVariantStyles = (variant: VariantType = 'default', theme?: Theme
...
+      width: '100%',
     },
-    '[data-component="leadingIcon"]': {
-      gridArea: 'leadingIcon',
+    '[data-component="leadingVisual"]': {
+      gridArea: 'leadingVisual',
     },
     '[data-component="text"]': {
       gridArea: 'text',
+      lineHeight: 'calc(20/14)',
```
Comment (mperrotti) [inline src/Button/styles.ts:277]: This is a much nicer way to read line heights! 😍

**Review by mperrotti — APPROVED** [review]
Looks great! The only thing I feel somewhat strongly about is that `trailingAction` is a strange prop name because it sounds like a button within a button.

Otherwise... :shipit:

File: .changeset/angry-spoons-grin.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (langermank) [reply .changeset/angry-spoons-grin.md:8]: We haven't implemented anything to manage overflow, its just a pattern recommendation to keep labels succinct. I'm not opposed to adding an ellipses or something. 

do and dont image showing a correct button with a single line label, and a multi line label in the "dont" example

File: src/Button/types.ts (hunk 7 above)
Comment (langermank) [reply src/Button/types.ts:60]: This one was [discussed](https://github.com/github/primer/discussions/1016) awhile ago here, and is also implemented already in PVC. My PR for the next major release will change `trailingIcon` to `trailingVisual` as well, which I think will better align with `trailingAction`. It's also part of the ActionList API which hasn't fully made its way into PRC.

File: src/Button/styles.ts (hunk 12 above)
Comment (langermank) [reply src/Button/styles.ts:283]: You can thank Vini for that one ❤️
